File: ios/AtmosFlowWidget/AmbientLayer.swift
import SwiftUI

/// The design's ambient sky, ported from the CSS the prototype draws it in.
///
/// Every tile in `AtmosFlow Widgets.dc.html` is the same recipe: a
/// time-of-day gradient, an optional dusk wash, a condition veil, and a
/// handful of absolutely-positioned shapes. The prototype writes those shapes
/// as inline CSS, so this is a small, deliberate subset of the CSS box model —
/// enough to reproduce them exactly, and nothing more.
///
/// Lengths come in three flavours because the design uses all three: a
/// fraction of the tile's reference size (most geometry, so a shape keeps its
/// proportions across the three widget sizes), a percentage of the box (used
/// for horizontal scatter), and a literal point value (hairlines — a rain
/// streak is 2pt wide at every size, exactly as in the design).
enum Length: Equatable {
    case frac(CGFloat)
    case pct(CGFloat)
    case pt(CGFloat)

    func resolve(against extent: CGFloat, scale: CGFloat) -> CGFloat {
        switch self {
        case .frac(let f): return f * scale
        case .pct(let p): return p * extent
        case .pt(let p): return p
        }
    }
}

/// The CSS box: any of the six offsets, resolved the way absolute positioning
/// resolves them — `left` and `width` win, `right` fills in when `left` is
/// absent, and the margins recentre a shape pinned to 50%.
struct Box: Equatable {
    var top: Length?
    var left: Length?
    var right: Length?
    var bottom: Length?
    var width: Length?
    var height: Length?
    var marginTop: Length?
    var marginLeft: Length?

    func rect(in size: CGSize, scale: CGFloat) -> CGRect {
        let w = width?.resolve(against: size.width, scale: scale) ?? 0
        let h = height?.resolve(against: size.height, scale: scale) ?? 0

        var x: CGFloat = 0
        if let left { x = left.resolve(against: size.width, scale: scale) }
        else if let right { x = size.width - right.resolve(against: size.width, scale: scale) - w }
        if let marginLeft { x += marginLeft.resolve(against: size.width, scale: scale) }

        var y: CGFloat = 0
        if let top { y = top.resolve(against: size.height, scale: scale) }
        else if let bottom { y = size.height - bottom.resolve(against: size.height, scale: scale) - h }
        if let marginTop { y += marginTop.resolve(against: size.height, scale: scale) }

        return CGRect(x: x, y: y, width: w, height: h)
    }
}

/// One of the design's keyframe animations, with this shape's own timing.
///
/// WidgetKit renders a still frame — a widget has no run loop to animate on —
/// so this is not played. It is sampled: `phase` gives the point in the cycle
/// a shape sits at, and because every drop and every star carries a different
/// duration and delay, sampling them all at one instant scatters them exactly
/// the way watching the prototype for a moment would.
struct Anim: Equatable {
    var name: String
    var duration: Double
    var delay: Double

    /// Where in its cycle this shape is at `t` seconds.
    func phase(at t: Double) -> Double {
        guard duration > 0 else { return 0 }
        let p = (t - delay).truncatingRemainder(dividingBy: duration) / duration
        return p < 0 ? p + 1 : p
    }
}

struct Glow: Equatable {
    var blur: Length
    var spread: Length
    var color: Color
}

struct MistBlob: Equatable {
    var width: CGFloat
    var height: CGFloat
    var x: CGFloat
    var y: CGFloat
    var color: Color
}

struct Stop: Equatable {
    var color: Color
    var at: CGFloat?
}

enum AmbientLayer: Equatable {
    /// A full-bleed wash — the condition veil, and the storm's lightning.
    case veil(Color, anim: Anim? = nil)
    /// The dusk gradient across the bottom of a dawn or evening sky.
    case duskVeil(height: CGFloat, color: Color)
    case radial(Box, center: UnitPoint, stops: [Stop], anim: Anim?)
    case ring(Box, strokeWidth: CGFloat, color: Color, anim: Anim?)
    case fill(Box, round: Bool, color: Color, glow: Glow?, anim: Anim?)
    case mist([MistBlob], anim: Anim?)
}

File: android/app/src/main/kotlin/com/surajshetty/atmos_flow/widget/AmbientLayer.kt
package com.surajshetty.atmos_flow.widget

/**
 * The design's ambient sky, in the same shape the iOS extension reads it in.
 *
 * Both platforms draw from one model — `tool/widget_spec/ambient_model.json`,
 * which is a reading of the CSS in the AtmosFlow Widgets design — so a tile
 * cannot drift between them without the design moving first.
 *
 * Lengths come in three flavours because the design uses all three: a fraction
 * of the tile's reference size, a percentage of the box, and a literal
 * dimension (hairlines — a rain streak is 2dp wide at every size).
 */
sealed interface Length {
    fun resolve(extent: Float, scale: Float): Float

    data class Frac(val value: Float) : Length {
        override fun resolve(extent: Float, scale: Float) = value * scale
    }

    data class Pct(val value: Float) : Length {
        override fun resolve(extent: Float, scale: Float) = value * extent
    }

    /** Device-independent pixels, scaled to the display like everything else. */
    data class Dp(val value: Float) : Length {
        override fun resolve(extent: Float, scale: Float) = value * Density.current
    }
}

/** The display's density, set once before a tile is drawn. */
object Density {
    var current: Float = 1f
}

/** The CSS box, resolved the way absolute positioning resolves it. */
data class Box(
    val top: Length? = null,
    val left: Length? = null,
    val right: Length? = null,
    val bottom: Length? = null,
    val width: Length? = null,
    val height: Length? = null,
    val marginTop: Length? = null,
    val marginLeft: Length? = null,
) {
    fun rect(w: Float, h: Float, scale: Float): FloatArray {
        val bw = width?.resolve(w, scale) ?: 0f
        val bh = height?.resolve(h, scale) ?: 0f

        var x = when {
            left != null -> left.resolve(w, scale)
            right != null -> w - right.resolve(w, scale) - bw
            else -> 0f
        }
        marginLeft?.let { x += it.resolve(w, scale) }

        var y = when {
            top != null -> top.resolve(h, scale)
            bottom != null -> h - bottom.resolve(h, scale) - bh
            else -> 0f
        }
        marginTop?.let { y += it.resolve(h, scale) }

        return floatArrayOf(x, y, x + bw, y + bh)
    }
}

/**
 * One of the design's keyframe animations, with this shape's own timing.
 *
 * An app widget draws into a [android.widget.RemoteViews] that the launcher
 * hosts — there is no run loop to animate on, so this is never played. It is
 * sampled, at [Ambient.INSTANT], exactly as the iOS extension samples it.
 */
data class Anim(val name: String, val duration: Float, val delay: Float) {
    fun phase(t: Float): Float {
        if (duration <= 0f) return 0f
        val p = ((t - delay) % duration) / duration
        return if (p < 0) p + 1 else p
    }
}

data class Glow(val blur: Length, val spread: Length, val color: Int)

data class MistBlob(
    val width: Float,
    val height: Float,
    val x: Float,
    val y: Float,
    val color: Int,
)

data class Stop(val color: Int, val at: Float? = null)

sealed interface AmbientLayer {
    val anim: Anim?

    /** A full-bleed wash — the condition veil, and the storm's lightning. */
    data class Veil(val color: Int, override val anim: Anim? = null) : AmbientLayer

    /** The dusk gradient across the bottom of a dawn or evening sky. */
    data class DuskVeil(val height: Float, val color: Int) : AmbientLayer {
        override val anim: Anim? get() = null
    }

    data class Radial(
        val box: Box,
        val centerX: Float,
        val centerY: Float,
        val stops: List<Stop>,
        override val anim: Anim?,
    ) : AmbientLayer

    data class Ring(
        val box: Box,
        val strokeWidth: Float,
        val color: Int,
        override val anim: Anim?,
    ) : AmbientLayer

    data class Fill(
        val box: Box,
        val round: Boolean,
        val color: Int,
        val glow: Glow?,
        override val anim: Anim?,
    ) : AmbientLayer

    data class Mist(
        val blobs: List<MistBlob>,
        override val anim: Anim?,
    ) : AmbientLayer
}
